File: laborations/mini/haskell/Compiler.hs
module Compiler where

import Control.Monad
import Control.Monad.State
import System.Environment (getArgs)
import System.Exit (exitFailure)

import AbsMini
import LexMini
import ParMini
import PrintMini
import ErrM

-- a simple-minded compiler that doesn't need type annotations and works for integers only

compile :: String -> Program -> String
compile name p = unlines $ reverse $ code $ execState (compileProgram name p) emptyEnv

compileProgram :: String -> Program -> State Env ()
compileProgram name (Prog stms) = do
  mapM_ emit [
    ".class public " ++ name,
    ".super java/lang/Object",
    "",
    ".method public <init>()V",
    "aload_0",
    "invokenonvirtual java/lang/Object/<init>()V",
    "return",
    ".end method",
    "",
    ".method public static main([Ljava/lang/String;)V",
    ".limit locals 100",  --- bogus limit
    ".limit stack 1000"   --- bogus limit
   ]
  mapM_ compileStm stms
  emit "return"
  emit ".end method"

compileStm :: Stm -> State Env ()
compileStm s = case s of
  SDecl t x   -> addVar x t
  SAss x e -> do
    compileExp e
    a <- lookupVar x
    emit ("istore " ++ show a) 
  SBlock stms -> do
    a <- newBlock
    mapM compileStm stms
    exitBlock a
  SPrint e -> do
    compileExp e
    emit $ "invokestatic Runtime/printInt(I)V"

compileExp :: Exp -> State Env ()
compileExp e = case e of
  EVar x  -> do
    a <- lookupVar x
    emit ("iload " ++ show a)
  EInt i    -> emit ("bipush " ++ show i)
  EDouble d -> emit ("ldc2_w " ++ show d)
  EAdd e1 e2 -> do
    compileExp e1
    compileExp e2
    emit "iadd"
  ETyped _ e -> compileExp e

data Env = E {
  addresses   :: [[(Ident,Address)]],
  nextLabel   :: Int,
  nextAddress :: Address,
  maxAddress  :: Address,
  stackSize   :: Int,
  maxSize     :: Int,
  code        :: [Instruction]
  }

emptyEnv :: Env
emptyEnv = E {
  addresses = [[]],
  nextLabel = 0,
  nextAddress = 1,
  maxAddress = 1,
  stackSize = 0,
  maxSize = 1,
  code = []
  }

type Instruction = String
type Address = Int

emit :: Instruction -> State Env ()
emit i = modify (\env -> env{code = i : code env})

addVar :: Ident -> Type -> State Env ()
addVar x t = modify (\env -> env {
  addresses = case addresses env of (scope:rest) -> (((x,nextAddress env):scope):rest),
  nextAddress = nextAddress env + typeSize t
  })

typeSize :: Type -> Int
typeSize t = case t of
  TInt -> 1
  TDouble -> 2

lookupVar :: Ident -> State Env Address
lookupVar x = do
  env <- get
  return $ look (addresses env) x 
 where
   look [] x = error $ "Unknown variable " ++ printTree x ++ "."
   look (scope:rest) x = case lookup x scope of
     Nothing -> look rest x
     Just a  -> a

newBlock :: State Env Address
newBlock = do
  modify (\env -> env {addresses = [] : addresses env})
  env <- get
  return $ nextAddress env

exitBlock :: Address -> State Env ()
exitBlock a = modify (\env -> env {
   addresses = tail (addresses env),
   nextAddress = a
   })

File: laborations/old-2015-mini/java1.5/Compiler.java
import Mini.Absyn.*;

import java.util.HashMap;
import java.util.LinkedList;

// compiler that works with the ordinary type checker, without type annotation
// all values are assumed to be integers

public class Compiler {

    public void compile(Program p) {
	Prog prog = (Prog)p;
	Env env = new Env();

	System.out.println(".class public Foo") ;  // should take class name from file
	System.out.println(".super java/lang/Object") ;
	System.out.println("") ;
	System.out.println(".method public <init>()V") ;
	System.out.println("aload_0") ;
	System.out.println("invokenonvirtual java/lang/Object/<init>()V") ;
	System.out.println("return") ;
	System.out.println(".end method") ;
	System.out.println("") ;
	System.out.println(".method public static main([Ljava/lang/String;)V") ;
	System.out.println(".limit locals 100") ; //bogus limit
	System.out.println(".limit stack 1000") ; //bogus limit

	for (Stm s : prog.liststm_) { 
          compileStm(s, env);
	}

	System.out.println("return") ;
	System.out.println(".end method") ;

    }

    private static class Env { 
	private LinkedList<HashMap<String,Integer>> scopes;
        private Integer maxvar;

	public Env() {
	    scopes = new LinkedList<HashMap<String,Integer>>();
            maxvar = 0 ;
	    enterScope();
	}

	public Integer lookupVar(String x) {
	    for (HashMap<String,Integer> scope : scopes) {
		Integer v = scope.get(x);
		if (v != null)
		    return v;
	    }
	    throw new RuntimeException("Unknown variable " + x + " in " + scopes);
	}

	public void addVar(String x) {
	    scopes.getFirst().put(x,maxvar);
            maxvar++;
	}

	public void enterScope() {
	    scopes.addFirst(new HashMap<String,Integer>());
	}

	public void leaveScope() {
	    scopes.removeFirst();
	}
    }

    private Env eenv = new Env() ;

    private void compileStm(Stm st, Object arg) {
	st.accept(new StmCompiler(), arg);
    }

    private class StmCompiler implements Stm.Visitor<Object,Object> {
	public Object visit(Mini.Absyn.SDecl p, Object arg) {
	    eenv.addVar(p.ident_);
	    return null;
	}

	public Object visit(Mini.Absyn.SAss p, Object arg) {
	    compileExp(p.exp_, arg);
            System.out.println("istore " + eenv.lookupVar(p.ident_));
	    return null;
	}

	public Object visit(Mini.Absyn.SBlock p, Object arg) {
            Integer oldmax = eenv.maxvar ;
	    eenv.enterScope();
	    for (Stm st : p.liststm_) {
		compileStm(st, arg);
	    }
	    eenv.leaveScope();
            eenv.maxvar = oldmax ;
	    return null;
	}

	public Object visit(Mini.Absyn.SPrint p, Object arg) {
	    compileExp(p.exp_, arg);
	    System.out.println("invokestatic Runtime/printInt(I)V") ;
	    return null;
	}
    }

    private Object compileExp(Exp e, Object arg) {
	return e.accept(new ExpCompiler(), arg);
    }

    private class ExpCompiler implements Exp.Visitor<Object,Object> {

	public Object visit(Mini.Absyn.EVar p, Object arg) {
	    System.out.println("iload " + eenv.lookupVar(p.ident_));
            return null ;
	}

	public Object visit(Mini.Absyn.EInt p, Object arg) {
	    System.out.println("bipush " + p.integer_);
            return null ;
	}

	public Object visit(Mini.Absyn.EDouble p, Object arg) {
	    System.out.println("ldc2_w " + p.double_);
            return null ;
	}

	public Object visit(Mini.Absyn.EAdd p, Object arg) {
	    compileExp(p.exp_1, arg) ;
	    p.exp_2.accept(this, arg) ;
	    System.out.println("iadd");
            return null ;
	}

	// used in the type annotating version if ETyped is in Absyn
	public Object visit(Mini.Absyn.ETyped p, Object arg) {
	    compileExp(p.exp_, arg) ;
            return null ;
	}
    }

}
